pkg/dab/find_test.go

```diff
@@ -0,0 +1,211 @@
+package dab
+
+import (
+	"io/fs"
+	"math/rand"
+	"path/filepath"
+	"testing"
+	"testing/fstest"
+
+	qt "github.com/frankban/quicktest"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/codec/json"
+	"github.com/warptools/warpforge/wfapi"
+)
+
+func MustMarshal(i interface{}, typeName string) []byte {
+	data, err := ipld.Marshal(json.Encode, i, wfapi.TypeSystem.TypeByName(typeName))
+	if err != nil {
+		panic(err)
+	}
+	return data
+}
+
+// Data blocks required for "GuessDocumentType"
+var (
+	ModuleData  = MustMarshal(&wfapi.ModuleCapsule{Module: &wfapi.Module{Name: "hello"}}, "ModuleCapsule")
+	PlotData    = MustMarshal(&wfapi.PlotCapsule{Plot: &wfapi.Plot{}}, "PlotCapsule")
+	FormulaData = MustMarshal(&wfapi.Formula{Action: wfapi.Action{Echo: &wfapi.Action_Echo{}}}, "Formula")
+)
+
+const Mode = 0444
+
+// randIrregular returns a fs.FileMode from modes in fs.ModeType except for ModeSymlink because stat/open will follow links
+func randIrregular() fs.FileMode {
+	modes := []fs.FileMode{fs.ModeDir, fs.ModeSymlink, fs.ModeNamedPipe, fs.ModeSocket, fs.ModeDevice, fs.ModeCharDevice, fs.ModeIrregular}
+	n := rand.Intn(len(modes))
+	return modes[n]
+}
+
+func TestFindActionableFromFS_NotFound(t *testing.T) {
+	t.Run("when no actionable files exist", func(t *testing.T) {
+		fsys := fstest.MapFS{}
+		m, p, f, path, rem, err := FindActionableFromFS(fsys, "", "non/existing/directory", true, ActionableSearch_Any)
+		qt.Assert(t, err, qt.IsNil)
+		qt.Assert(t, m, qt.IsNil)
+		qt.Assert(t, p, qt.IsNil)
+		qt.Assert(t, f, qt.IsNil)
+		qt.Assert(t, path, qt.Equals, "")
+		qt.Assert(t, rem, qt.Equals, "")
```

Comment:
Reading these test cases makes me wonder if there really should be a struct here.

Either a struct that we plonk all the function returns into (just so we can assert everything at once, and to help condense the tests into a table)....

... or this is really making me wonder if I was wrong for not just having this function return a struct in the first place.

Reply:
Hm.  Reading further test cases below, now I see that having the function return a struct wouldn't solve both desires at once, because the later tests are using `qt.IsNotNil`.

It might be an option to make a struct of bools and have a helper function that reduces return value nillity to that struct of bools.  That would help with the condensing into a table.  Maybe still worth it?

Reply:
We can make condense some of it but doing tables with so many diverse inputs and outputs is going to be pretty rough looking regardless of what we do.